### PR TITLE
🔨 Better type for numeric mcdoc literals

### PR DESCRIPTION
**Pull request by NeunEinser**
This makes numeric literal types more expressive in their type.

This makes checking these types against nodes a bit less cumbersome in my current related work.

### CODE REVIEW

File: packages/mcdoc/src/binder/index.ts
Comment (SPGoding) [inline packages/mcdoc/src/binder/index.ts:1233]: Can we use `NumericTypeIntKind` instead of `(typeof NumericTypeIntKinds)[number]`?

**Comment by NeunEinser** [comment]
Looks like there is a failing test now, because of the node type change. I'll wait until I have feedback if this is a desirable change until looking into how to fix those.

File: packages/mcdoc/src/binder/index.ts
Comment (SPGoding) [inline packages/mcdoc/src/binder/index.ts:1219]: `Number.isInteger(1.0)` would be `true` so we would have an `int` literal number when it is in fact a floating point number. Probably have to change the spec and the parser to properly fix this (the spec current specifies `float` + suffix instead of using `float` and `interger` properly cuz i was too lazy).

**Comment by SPGoding** [comment]
Otherwise I think this would be a good change.

File: packages/mcdoc/src/binder/index.ts
Comment (NeunEinser) [reply packages/mcdoc/src/binder/index.ts]: Yep, that's less cursed for sure

File: packages/mcdoc/src/binder/index.ts
Comment (NeunEinser) [reply packages/mcdoc/src/binder/index.ts:1219]: Hmm, good point

Comment (NeunEinser) [reply packages/mcdoc/src/binder/index.ts]: I changed the parser to emit an integer node when there is no dot and no float type marker

**Comment by NeunEinser** [comment]
I can't figure out how to fix the tests. From the output it looks like it tries to see if the current mcdoc exports are equivalent to what this branch would generate, but there is ofc an intentional mismatch. Not sure how to change the test data.

**Comment by SPGoding** [comment]
You can set the environment variable `SNAPSHOT_UPDATE=1` when running `npm test` or just delete the `__snapshots__` folder and let `npm test` re-generate that. We should probably add this to README

**Comment by NeunEinser** [comment]
It deleted `__snapshots__/packages/core/test-out/service/Operations.spec.js`, not sure why.

**Comment by SPGoding** [comment]
> It deleted `__snapshots__/packages/core/test-out/service/Operations.spec.js`, not sure why.

the source file was moved from `service/` to `common/` so that's just an orphan snapshot.

**Review by SPGoding — APPROVED** [review]
LGTM, would be nice if you can update the syntax for `TYPED_NUMBER` in the mcdoc spec (`docs/user/mcdoc/index.adoc`) as well. Something like this probably works but it's been a while since I've worked on that punctuation soup document:

```adoc
<<t-typed-number>>::
	<<t-float>> ++[++`d` `D` `f` `F`++]++^?^ | +
	<<t-integer>> ++[++`b` `B` `l` `L` `s` `S`++]++^?^
```

**Comment by NeunEinser** [comment]
(rebase, no new commits)

**Comment by TheAfroOfDoom** [comment]
requested myself as a non-blocking reviewer so i remember to look at this eventually

File: packages/mcdoc/src/binder/index.ts
Comment (misode) [inline packages/mcdoc/src/binder/index.ts:92]: Unused import

**Review by misode — APPROVED** [review]

File: __snapshots__/packages/core/test-out/service/StateProxy.spec.js
Comment (TheAfroOfDoom) [inline __snapshots__/packages/core/test-out/service/StateProxy.spec.js:1]: i might prefer we move these file deletions into a separate PR that removes the orphaned snapshots? as-is in here it's almost implying that these mcdoc type changes caused these file diffs

Comment (TheAfroOfDoom) [reply __snapshots__/packages/core/test-out/service/StateProxy.spec.js:1]: maybe more leg-work than is worth though

File: packages/mcdoc/src/type/index.ts
Comment (TheAfroOfDoom) [inline packages/mcdoc/src/type/index.ts:153]: this is really nice

File: __snapshots__/packages/mcdoc/test-out/__fixture__/struct/simple.spec.js
Comment (TheAfroOfDoom) [inline __snapshots__/packages/mcdoc/test-out/__fixture__/struct/simple.spec.js:51]: is there any meaningful difference in this diff? does it matter that we parse release versions as a "double" rather than just some number?

we should probably parse it as a "semantic version" type instead, but that's prob better for a later PR

**Review by TheAfroOfDoom — COMMENTED** [review]
couple smaller things. generally this seems like a good idea though, nice work 😸 

---

@SPGoding what are your thoughts on removing `/__snapshots__/** linguist-generated=true` from `.gitattributes`?

i guess they technically _are_ generated, but it'd be nice if they weren't auto not rendered in the `Files changed` view. i find myself definitely wanting to review each snapshot change for robustness

---

similarly, why is `README.md` set as `linguist-generated=true`?

File: __snapshots__/packages/mcdoc/test-out/__fixture__/struct/simple.spec.js
Comment (NeunEinser) [reply __snapshots__/packages/mcdoc/test-out/__fixture__/struct/simple.spec.js:51]: It's parsed as string when it's `1.19.3`.

According to mcdoc spec, when not specifying a number suffix, it is parsed as either an int or a double, depending on its value.

File: __snapshots__/packages/core/test-out/service/StateProxy.spec.js
Comment (NeunEinser) [reply __snapshots__/packages/core/test-out/service/StateProxy.spec.js:1]: I could revert the changes to these deleted files, just not sure if it's wanted.

**Comment by SPGoding** [comment]
> what are your thoughts on removing /__snapshots__/** linguist-generated=true from .gitattributes?

i agree

> similarly, why is README.md set as linguist-generated=true?

it used to have automatically generated contributor information. i guess we can remove this as well.

File: __snapshots__/packages/core/test-out/service/StateProxy.spec.js
Comment (TheAfroOfDoom) [reply __snapshots__/packages/core/test-out/service/StateProxy.spec.js:1]: eh its probably fine to keep as-is. would be nice if we eventually had a script to check for orphaned snapshots in GHA

**Review by TheAfroOfDoom — APPROVED** [review]

**Pull request merged**
